### PR TITLE
Add CLOGS Integration

### PR DESCRIPTION
**Pull request by parc-jason**
# Description
Remove direct access from CDOGS to logstash.
Added common-logging library code to deliver logging information and api stats tracking to CLOGS through its API.
Added unit tests for coverage in the common-logging code.
Common Logging code has a queue (in memory) and a timer to flush the queue.  The batch size and interval is set in config map.
There is a separate service client (CDOGS_SERVICE_CLIENT) for calling CLOGS.

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the OpenAPI 3.0 `v*.api-spec.yaml` documentation (if appropriate)
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-idcqvl-tools.pathfinder.gov.bc.ca/dashboard?id=common-document-generation-service-pr-25

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-idcqvl-tools.pathfinder.gov.bc.ca/dashboard?id=common-document-generation-service-pr-25

File: app/src/common-logging/queue-memory.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  /* Returns the item at the front of the queue (without dequeuing it). If the
+   * queue is empty then undefined is returned.
+   */
+  peek() {
+    return (this.queue.length > 0 ? this.queue[this.offset] : undefined);
+  }
+
+
```
Comment (loneil) [inline app/src/common-logging/queue-memory.js]: Extra space

File: app/src/common-logging/README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+Accepts a message (string or JSON), and options (format, level, retention, parse).  
+The message parameter will be set the api object message field or data field (if message is JSON).  The options can be used to set corresponding api log object fields.  
+  
+*NOTE*: the parse option is a function that can parse the message parameter and turn it into JSON. In such a case, the api log object data field is set instead of message.  
+
+### queue (common-logging-queue)
+
+| option | description |
+| --- | --- |
+| batchSize | how many items to place in a single API call |
```
Comment (jujaga) [inline app/src/common-logging/README.md]: perhaps "maxBatchSize" may be more descriptive?

File: app/src/common-logging/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+```
+
+
+
+
+ 
+  
+
+
```
Comment (jujaga) [inline app/src/common-logging/README.md]: Unnecessary whitespace lines

File: Jenkinsfile [hunk 4]
```diff
@@ -260,7 +260,9 @@ def deployStage(String stageEnv, String projectEnv, String hostRouteEnv) {
       echo "Checking for ConfigMaps and Secrets in project ${openshift.project()}..."
       if(!(openshift.selector('cm', "cdogs-keycloak-config").exists() &&
       openshift.selector('cm', "cdogs-server-config").exists() &&
-      openshift.selector('secret', "cdogs-keycloak-secret").exists())) {
+      openshift.selector('cm', "cdogs-clogs-config").exists() &&
+      openshift.selector('secret', "cdogs-keycloak-secret").exists() &&
+      openshift.selector('secret', "cdogs-clogs-secret").exists())) {
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply Jenkinsfile]: OK, yeah, good point.  I've changed to cdogs-common-service-*

File: app/src/common-logging/common-logging-http.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    this.axios = this.connection.axios;
+    this.apiUrl = options.apiUrl;
+  }
+
+  async xfer(messages) {
+    if (messages) {
+      const batch = Array.isArray(messages) ? messages : [messages];
+
+      if (!batch.length) {
+        return -2;
```
Comment (jujaga) [inline app/src/common-logging/common-logging-http.js]: It might not be a bad idea to have these negative number values be specified in a constants file - otherwise they're just potentially magic numbers without too much context as to what they represent.

File: app/src/components/commonlogger.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+const config = require('config');
+
+const CommonLogger = require('../common-logging/common-logger');
+const CommonLoggingHttp = require('../common-logging/common-logging-http');
+const CommonLoggingQueue = require('../common-logging/common-logging-queue');
+const CommonLoggingStdout = require('../common-logging/common-logging-stdout');
```
Comment (jujaga) [inline app/src/components/commonlogger.js]: For future consideration, we may want to have the common-logging have a top-level index.js which would require the above classes instead. That way, a user only has to import the library instead of having to refer to 4 discrete requires.

File: openshift/app.bc.yaml [hunk 7]
```diff
@@ -48,6 +48,7 @@ objects:
         dockerfile: |-
           FROM BuildConfig
           ARG APP_ROOT=/opt/app-root/src
+          ENV NO_UPDATE_NOTIFIER=true
           ENV PATH="/usr/lib/libreoffice/program:${PATH}" \
               PYTHONUNBUFFERED=1
```
Comment (jujaga) [inline openshift/app.bc.yaml]: Didn't realize this already had an ENV layer. We can reduce one docker layer in the build if we combined the lines together
```
ENV NO_UPDATE_NOTIFIER=true \
    PATH="/usr/lib/libreoffice/program:${PATH}" \
    PYTHONUNBUFFERED=1
```

File: app/src/common-logging/common-logging-http.js (hunk 5 above)
Comment (parc-jason) [reply app/src/common-logging/common-logging-http.js]: They are really only there for testing purposes.
We can deal when/if we move to a npm package.

File: app/src/components/commonlogger.js (hunk 6 above)
Comment (parc-jason) [reply app/src/components/commonlogger.js]: Yup, if/when we move to npm package, we will have to consider what is the easiest way for a consumer.

File: openshift/app.bc.yaml (hunk 7 above)
Comment (parc-jason) [reply openshift/app.bc.yaml]: OK, updated.
